### DISCUSSION TITLE
Renovate ignores private NPM registry to download Yarn, corepack related

### DISCUSSION THREAD

**Discussion by dirkluijk** [renovatebot/renovate#26615] [Request Help] [answered]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us what version of Renovate you run.

37.127.0

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### Was this something which used to work for you, and then stopped?

It used to work, and then stopped

### Describe the problem

We are running Renovate in a Gitlab CI pipeline on a self-hosted Gitlab instance which - by company policy - does not allow internet connections. We have our own NPM registry (Artifactory) which is accessible and correctly configured using an `.npmrc` file.

For some reason, Renovate recently stopped working. It has issues, related to corepack, resulting in an "Artifact update problem".

Before this issue, I didn't even know what corepack was. We didn't change anything in our repository regarding node/yarn AFAIK.

See attached logs.

### Relevant debug logs

<details><summary>Logs</summary>

```

{
           "lockFile": "*****/yarn.lock",
           "stderr": "Internal Error: Error when performing the request to https://registry.npmjs.org/yarn; for troubleshooting help, see https://github.com/nodejs/corepack#troubleshooting\n    at ClientRequest.<anonymous> (/opt/containerbase/tools/corepack/0.24.0/node_modules/corepack/dist/lib/corepack.cjs:42208:16)\n    at ClientRequest.emit (node:events:517:28)\n    at TLSSocket.socketErrorListener (node:_http_client:501:9)\n    at TLSSocket.emit (node:events:517:28)\n    at emitErrorNT (node:internal/streams/destroy:151:8)\n    at emitErrorCloseNT (node:internal/streams/destroy:116:3)\n    at process.processTicksAndRejections (node:internal/process/task_queues:[82](https://gitlab.****.io/******/-/jobs/16648525#L82):21)\n"
}
```

</details>

### Have you created a minimal reproduction repository?

I have explained in the description why a minimal reproduction is impossible

**Comment by rarkins** [comment]
Does your CI set `COREPACK_NPM_REGISTRY`? I think that controls this and not `.npmrc`. In fact I thought that Yarn 2+ does not read `.npmrc`, and isn't installable via npm either - so I would be surprised if it's used for controlling the corepack installation of Yarn

**Reply by dirkluijk** [reply]
Hi @rarkins , no we don't set `COREPACK_NPM_REGISTRY` as far as I can tell. As far as I can see, it's not running Yarn yet; but instead it's corepack trying to download Yarn in the first place. So it seems corepack doesn't pickup `.npmrc` when downloading Yarn from the NPM registry.

But regardless of that, I don't want to use corepack at all. Yarn should be pre-installed on the renovate image, from what I've been told.

Can you confirm that Renovate uses Corepack? Is there a way to explicitly disable this?

**Reply by rarkins** [reply]
Are you using the "37.127.0-full" tag of Renovate? Renovate does not have preinstalled managers in the default image tags.

And which version of Yarn are you using?

**Reply by dirkluijk** [reply]
We are using Yarn 1.x (using Maven Frontend Plugin) in our build, but Renovate installs its own version of Yarn.

We are using `renovate/renovate:37-full` as base image.

**Reply by dirkluijk** [reply]
It seems Yarn is available.
```
$ docker run renovate/renovate:37-full yarn -v
1.22.21
```

Interestingly,

```
$ cat /usr/local/bin/yarn
#!/usr/bin/env node
require('./lib/corepack.cjs').runMain(['yarn', ...process.argv.slice(2)]);
```
So it does seem to use corepack, which is not our intention.

**Reply by breun** [reply]
According to https://github.com/renovatebot/renovate/issues/14851 corepack was intentionally not enabled in Renovate. It's also still marked as an experimental feature, and doesn't seem to support downloading a package manager from an alternative location yet, as far as I can tell, so that's a problem in environments that don't have public internet access.

**Reply by rarkins** [reply]
That's a very long time ago and not the right image

**Reply by dirkluijk** [reply]
I will see if I can reproduce the issue in a public repository, if I can disable internet access somehow in a Gitlab action or something.

**Reply by breun** [reply]
> That's a very long time ago and not the right image

This is current (image built 8 hours ago):

```
❯ docker run -it --rm renovate/renovate:37-full bash
ubuntu@cf83ee59f31a:/usr/src/app$ renovate --version
37.129.1
ubuntu@cf83ee59f31a:/usr/src/app$ which yarn
/usr/local/bin/yarn
ubuntu@cf83ee59f31a:/usr/src/app$ cat /usr/local/bin/yarn
#!/bin/bash

if [[ -z "${CONTAINERBASE_ENV+x}" ]]; then
  . /usr/local/etc/env
fi
/opt/containerbase/tools/corepack/0.24.0/bin/yarn "$@"
ubuntu@cf83ee59f31a:/usr/src/app$ cat /opt/containerbase/tools/corepack/0.24.0/bin/yarn
#!/usr/bin/env node
require('./lib/corepack.cjs').runMain(['yarn', ...process.argv.slice(2)]);
```

So it indeed looks like the `yarn` binary that is included in a freshly built `renovate/renovate:37-full` image uses `corepack`.

**Reply by rarkins** [reply]
Yes, Renovate's images generally use the latest stable version of each tool, and the way to install Yarn these days is via corepack: https://yarnpkg.com/getting-started/install

So this is intentional/not a bug/etc. It just confirms that Renovate has a recent version of Yarn by default. We won't ship Renovate using a deprecated version of Yarn (v1) by default.

But Renovate can dynamically install Yarn v1 at runtime if you allow it:

```
17:05 $ docker run --rm -it renovate/renovate /bin/bash
Unable to find image 'renovate/renovate:latest' locally
latest: Pulling from renovate/renovate

Digest: sha256:9b7fe1ee100f0aedd3dda50e4366993a56068d618ac0b3bf4401f4925a2435ef
Status: Downloaded newer image for renovate/renovate:latest
ubuntu@dd2618d2e2a7:/usr/src/app$ which yarn
/usr/local/bin/yarn
ubuntu@dd2618d2e2a7:/usr/src/app$ install-tool yarn 1.22.19
[16:07:23.407] INFO (12): Installing npm yarn@1.22.19...
1.22.19
[16:07:24.061] INFO (12): Installed tool yarn in 654ms.
ubuntu@dd2618d2e2a7:/usr/src/app$ yarn --version
1.22.19
```

If your requirements are:
- I need to use Yarn v1
- I cannot install anything at runtime

Then you need to build a custom image, e.g. like

```
FROM renovate/renovate
RUN install-tool yarn 1.22.19
```

I think the corepack discussion above is kind of a red herring and a result of the problem, not the root cause itself. The root cause seems to be that you need a deprecated version of a tool installed by default, which we can't support.

**Reply by dirkluijk** [reply]
Ah yeah that makes sense. I guess the same reasoning applies for the NPM version used?

**Reply by breun** [reply]
In this case things can be installed at runtime, but the public internet is not accessible, only an internal artifact manager that can proxy things like NPM and Maven Central.

@dirkluijk I guess that setting `COREPACK_NPM_REGISTRY` to the URL of the internal NPM proxy could make this work.

**Comment by nilson-aguiar** [comment]
> Does your CI set `COREPACK_NPM_REGISTRY`? I think that controls this and not `.npmrc`. In fact I thought that Yarn 2+ does not read `.npmrc`, and isn't installable via npm either - so I would be surprised if it's used for controlling the corepack installation of Yarn

Hi @rarkins, 

I'm also going through the same problem, I've tried setting the `COREPACK_NPM_REGISTRY ` through the renovate.json file:

![image](https://github.com/renovatebot/renovate/assets/5804096/7cc813ea-c30e-4f57-b9f4-53c553b07a15)

and the env variable is not defined when running yarn install:

![image](https://github.com/renovatebot/renovate/assets/5804096/c4a779f1-da53-4d26-912c-e465d6522616)

do you know what am I doing wrong?

**Reply by rarkins** [reply]
I think we probably need to modify code to ensure it's passed through to such commands. You could try settinge exposeAllEnv=true ([docs](https://docs.renovatebot.com/self-hosted-configuration/#exposeallenv)) in the meantime

**Reply by nilson-aguiar** [reply]
shall I create a ticket for it?

**Reply by viceice** [reply]
`customEnvVariables`[^1] need to be in global bot config, not repo config.

[^1]: https://docs.renovatebot.com/self-hosted-configuration/#customenvvariables

**Reply by rarkins** [reply]
I think we should pass through this env by default, so that no extra renovate config is necessary

**Reply by rarkins** [reply, marked as answer]
https://github.com/renovatebot/renovate/pull/26664

**Reply by breun** [reply]
The corepack environment variables are now passed on, but it seems that [corepack doesn't support downloading Yarn from a configurable location yet](https://github.com/nodejs/corepack/issues/296), so Renovate still won't be able to download Yarn in an environment without public internet access.

**Reply by rarkins** [reply]
I guess the old answer still applies - if you can't support corepack installs from the internet, and you need to default to Yarn v1, then you need a custom image.

@viceice as an alternative, do you think we could extend install-tool to allow a custom binary name (such as "yarn1") and then we could install both yarn1 and yarn by default? Then Renovate could be enhanced with this knowledge if COREPACK=1 in env or whatever

**Reply by viceice** [reply]
I'm now confused. corepack isn't using the passed npm registry to download yarn v1?

you can install yarn v1 with `install-tool yarn 1.22.21`. I think renovate will already do it on dynamic install.

I think in this case the user should extend our docker image and install yarn v1 with the command above.

**Reply by rarkins** [reply]
Yarn 2+ is not installed via npm registry, so it isn't included

**Reply by viceice** [reply]
https://github.com/nodejs/corepack/blob/a3f271cdf04f652fcc63547705ef0fb95f593c2c/config.json#L140

you are right 🤔

we're preparing yarn v1 on renovate images

https://github.com/renovatebot/base-image/blob/4bb822a5d5368e55bf2b3419aa51033fe2f66e04/Dockerfile#L30

container base doesn't support to install yarn v2+. we never implemented it, because they suggest to use corepack or add a repo local yarn binary.

**Reply by viceice** [reply]
🤔  we're installing yarn v1 by default on our renovate images

https://github.com/renovatebot/base-image/blob/45e48abaa92837bafd9efbf8cbbbf8220cdb77ad/Dockerfile#L19-L33
